File: scripts/run_circuit_bound_proof.py
#!/usr/bin/env python3
"""Proofs assembled from the shares the circuit actually wrote.

The quote proof is assembled from shares. Until the circuit kept them, those
shares reached the prover by a route of their own: the circuit computed, the
prover was handed values, and nothing said the two were the same numbers. A
proof about numbers that merely agree with a computation is not a proof about
the computation.

`test_share_binding` closed that for one value, the winner. This closes it for
the wires the proof is actually made of. The circuit is run with
`--persist-wires --shamir-inputs`, so every node keeps its share of every wire
*in the field the commitments use*, and the proofs below are built from those
files and nothing else.

What is still supplied from outside, and honestly cannot come from the circuit:

* **Blindings.** A Pedersen blinding is not something the computation knows
  about. Registered fields carry what the maker shared at registration; derived
  wires carry what the nodes generated between them. Neither is a hole --- the
  commitment is computed *from the shares* here, so no wire is opened to commit
  to it.
* **Cross terms.** A product proof needs `s = r_c - r_a*b`, a product of two
  secrets. In a deployment that is one multiplication, which is what the
  protocol already runs. Here the harness computes it, and that is the one place
  this script reconstructs anything --- marked, because it is the boundary.
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from mp_spdz.persistence import WIRE_NAMES, read_wires, reconstruct  # noqa: E402
from scripts.hosts import this_host                                  # noqa: E402
from zk.commit import Pedersen, verify_product                       # noqa: E402
from zk.groups import make_group                                     # noqa: E402
from zk.threshold_gadgets import (Shared, commitment_from_shares,    # noqa: E402
                                  joint_prove_bit, joint_prove_product,
                                  verify_square_bit)
from zk.threshold_quote import _share                                # noqa: E402


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--persistence", type=Path, required=True)
    ap.add_argument("--out", type=Path,
                    default=ROOT / "artifacts" / "circuit_bound_proof.json")
    ap.add_argument("--parties", type=int, default=7)
    ap.add_argument("--threshold", type=int, default=2)
    ap.add_argument("--makers", type=int, default=4)
    ap.add_argument("--qty", type=int, default=100)
    args = ap.parse_args()

    group = make_group("ed25519")
    key = Pedersen(group, b"qomm:policy:v1")
    parties = list(range(1, args.parties + 1))
    quorum = parties[: args.threshold + 1]

    wires = read_wires(args.persistence, args.parties, args.makers)
    if wires["prime"] != group.order:
        raise SystemExit(
            f"the circuit wrote in a {wires['prime'].bit_length()}-bit field and "
            f"the commitments use a {group.order.bit_length()}-bit one; run it "
            "with --shamir-inputs")

    def opened(shares, subset=None):
        """Only the harness does this, and only to stand in for a multiplication."""
        subset = subset or quorum
        return reconstruct([(p, shares[p]) for p in subset], wires["prime"])

    def wire(shares) -> tuple[Shared, dict]:
        blinding = _share(group, group.random_scalar(), parties, args.threshold)
        return Shared(commitment_from_shares(key, shares, blinding, quorum),
                      dict(shares), blinding), blinding

    rows = []
    for index, maker in enumerate(wires["makers"]):
        built = {name: wire(maker[name]) for name in WIRE_NAMES}
        shared = {name: built[name][0] for name in WIRE_NAMES}
        blind = {name: built[name][1] for name in WIRE_NAMES}

        # every commitment came from shares; check it against the direct one
        agrees = all(
            group.encode(shared[name].commitment)
            == group.encode(key.commit(opened(maker[name]),
                                       opened(blind[name])))
            for name in WIRE_NAMES)

        # depth = slope * qty, on the circuit's own slope and depth shares
        r_slope, r_depth = opened(blind["slope"]), opened(blind["depth"])
        qty_blinding = _share(group, group.random_scalar(), parties, args.threshold)
        qty_shared = Shared(
            commitment_from_shares(key, _share(group, args.qty, parties,
                                               args.threshold),
                                   qty_blinding, quorum),
            _share(group, args.qty, parties, args.threshold), qty_blinding)
        cross = _share(group, (r_depth - r_slope * args.qty) % group.order,
                       parties, args.threshold)
        depth_proof = joint_prove_product(
            key, shared["slope"].commitment, qty_shared, shared["depth"], cross,
            quorum, args.threshold, b"circuit:depth")
        depth_ok = verify_product(key, shared["slope"].commitment,
                                  qty_shared.commitment, shared["depth"].commitment,
                                  depth_proof, b"circuit:depth")

        # `ok` and `fits` are bits the circuit produced. Proving they are bits
        # is the square proof, and it is the step the disjunction could not do.
        bit_results = {}
        for name in ("fits", "ok", "active"):
            r = opened(blind[name])
            b = opened(maker[name])
            proof = joint_prove_bit(
                key, shared[name],
                _share(group, (r * (1 - b)) % group.order, parties, args.threshold),
                quorum, args.threshold, b"circuit:" + name.encode())
            bit_results[name] = verify_square_bit(
                key, shared[name].commitment, proof, b"circuit:" + name.encode())

        # The controls. A check that cannot fail is not a check, and every
        # result above is a boolean that came out True.
        #
        # The first control --- prove `depth` against a quantity the circuit did
        # not use --- only bites where the slope is non-zero. At `slope = 0` the
        # statement "depth = slope * anything" is *true*, both sides being zero,
        # and the proof is right to accept it. Two of the four makers in this
        # run have a zero slope, so two of them report the control as not
        # refusing, and that is arithmetic rather than a weak proof. Reporting
        # it as a pass would have been the lie; `control_applies` says which.
        slope_value = opened(maker["slope"])
        wrong_qty = args.qty + 1
        wrong_shares = _share(group, wrong_qty, parties, args.threshold)
        wrong_blinding = _share(group, group.random_scalar(), parties, args.threshold)
        wrong = Shared(commitment_from_shares(key, wrong_shares, wrong_blinding,
                                              quorum),
                       wrong_shares, wrong_blinding)
        wrong_cross = _share(group, (r_depth - r_slope * wrong_qty) % group.order,
                             parties, args.threshold)
        wrong_proof = joint_prove_product(
            key, shared["slope"].commitment, wrong, shared["depth"], wrong_cross,
            quorum, args.threshold, b"circuit:depth")
        refused = not verify_product(key, shared["slope"].commitment,
                                     wrong.commitment, shared["depth"].commitment,
                                     wrong_proof, b"circuit:depth")

        # A second control that bites everywhere: the same proof checked
        # against another maker's depth commitment. Nothing makes that true.
        other = (index + 1) % args.makers
        other_depth = commitment_from_shares(
            key, wires["makers"][other]["depth"],
            _share(group, group.random_scalar(), parties, args.threshold), quorum)
        crossed = not verify_product(key, shared["slope"].commitment,
                                     qty_shared.commitment, other_depth,
                                     depth_proof, b"circuit:depth")

        rows.append({"maker": index,
                     "slope": slope_value if slope_value < group.order // 2
                              else slope_value - group.order,
                     "wrong_quantity_control_applies": slope_value % group.order != 0,
                     "a_wrong_quantity_is_refused": refused,
                     "another_makers_depth_is_refused": crossed,
                     "commitments_from_shares_agree": agrees,
                     "depth_is_slope_times_qty": depth_ok,
                     "bits_proved": bit_results,
                     "no_node_holds_a_wire": all(
                         maker[name][p] != opened(maker[name])
                         for name in WIRE_NAMES for p in parties
                         if opened(maker[name]) > 1)})
        print(f"maker {index}: commitments {agrees}, depth {depth_ok}, "
              f"bits {bit_results}, wrong qty refused {refused} "
              f"(control applies: {slope_value % group.order != 0}), "
              f"other depth refused {crossed}", flush=True)

    payload = {
        "host": this_host(),
        "field_bits": wires["prime"].bit_length(),
        "field_matches_commitment_scalar_field": wires["prime"] == group.order,
        "parties": args.parties, "threshold": args.threshold,
        "wires_per_maker": len(WIRE_NAMES), "wire_order": list(WIRE_NAMES),
        "runs_in_file": wires["runs_in_file"],
        "winner_key": opened(wires["winner_key"]),
        "supplied_outside_the_circuit": [
            "blindings, which the computation does not know about",
            "cross terms, which are one multiplication in a deployment",
        ],
        "rows": rows,
    }
    args.out.parent.mkdir(parents=True, exist_ok=True)
    args.out.write_text(json.dumps(payload, indent=1) + "\n")
    print(f"wrote {args.out}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/qomm-harness/src/bin/run_circuit_bound_proof.rs
//!
//! The witness shares are read from MP-SPDZ persistence files. Run the circuit
//! with both `--persist-wires` and `--shamir-inputs`; otherwise there are no
//! circuit-owned wires in the ristretto255 scalar field to prove statements
//! about.

use curve25519_dalek::scalar::Scalar;
use merlin::Transcript;
use qomm_harness::{parse_value, repo_root, write_pretty_json, HarnessResult};
use qomm_mpc::persistence::{read_wires, FieldElement, Wires, WIRE_NAMES};
use rand::rngs::OsRng;
use serde_json::{json, Map};
use std::collections::BTreeMap;
use std::ffi::OsString;
use std::path::PathBuf;
use zkfmi_zk::pedersen::Pedersen;
use zkfmi_zk::shamir;
use zkfmi_zk::sigma::verify_product;
use zkpi_proofs::threshold_gadgets::{
    coefficient_commitments_from_evaluations, commitment_from_shares,
    joint_prove_product_from_contributions, verify_square_bit, ProductNodeContribution, Shared,
};
use zkpi_proofs::threshold_quote::RISTRETTO_SCALAR_ORDER_LE;
use zkpi_proofs::threshold_sigma::{PartyId, ScalarShares};

struct Options {
    persistence: PathBuf,
    out: PathBuf,
    parties: usize,
    threshold: usize,
    makers: usize,
    qty: i64,
}

fn main() {
    if let Err(error) = run_main() {
        eprintln!("{error}");
        std::process::exit(1);
    }
}

fn run_main() -> HarnessResult<()> {
    let options = parse_args()?;
    if options.parties <= options.threshold {
        return Err("--parties must be greater than --threshold".into());
    }
    if options.makers == 0 {
        return Err("--makers must be positive".into());
    }
    if options.qty < 0 {
        return Err("--qty must be non-negative".into());
    }

    let wires = read_wires(&options.persistence, options.parties, options.makers, -1)?;
    let prime_le: [u8; 32] = wires
        .prime
        .to_bytes_le(32)?
        .try_into()
        .expect("a requested 32-byte encoding");
    if prime_le != RISTRETTO_SCALAR_ORDER_LE {
        return Err(format!(
            "the circuit wrote in a {}-bit field and the commitments use a 253-bit one; run it with --persist-wires --shamir-inputs",
            wires.prime.bit_length()
        )
        .into());
    }

    let key = Pedersen::new(b"qomm:policy:v1");
    let parties = (1..=options.parties).collect::<Vec<_>>();
    let quorum = parties[..=options.threshold].to_vec();
    let mut rng = OsRng;
    let scalar_wires = scalar_wires(&wires)?;
    let winner_key = reconstruct(&scalar_wires.winner_key, &quorum)?;
    let mut rows = Vec::new();

    for (index, maker) in scalar_wires.makers.iter().enumerate() {
        let mut shared = BTreeMap::new();
        let mut blindings = BTreeMap::new();
        for name in WIRE_NAMES {
            let (wire, blinding) = wire_from_shares(
                &key,
                maker
                    .get(name)
                    .ok_or_else(|| format!("maker {index}: missing wire {name}"))?,
                &parties,
                &quorum,
                options.threshold,
                &mut rng,
            )?;
            shared.insert(name, wire);
            blindings.insert(name, blinding);
        }

        let commitments_from_shares_agree = WIRE_NAMES.iter().all(|name| {
            let value = reconstruct(&maker[name], &quorum).expect("validated share map");
            let blinding = reconstruct(&blindings[name], &quorum).expect("validated blinding map");
            shared[name].commitment.compress() == key.commit(&value, &blinding).compress()
        });

        let slope_blinding = reconstruct(&blindings["slope"], &quorum)?;
        let depth_blinding = reconstruct(&blindings["depth"], &quorum)?;
        let qty_value = scalar_from_i64(options.qty);
        let qty_shares = share_scalar(&qty_value, &parties, options.threshold, &mut rng);
        let qty_blinding_secret = Scalar::random(&mut rng);
        let qty_blindings =
            share_scalar(&qty_blinding_secret, &parties, options.threshold, &mut rng);
        let qty_shared =
            shared_from_maps(&key, qty_shares, qty_blindings, &quorum, options.threshold)?;
        let depth_cross_secret = depth_blinding - slope_blinding * qty_value;
        let depth_cross = share_scalar(&depth_cross_secret, &parties, options.threshold, &mut rng);
        let depth_contributions = node_product_contributions(&qty_shared, &depth_cross, &quorum)?;
        let mut transcript = tagged(b"circuit:depth");
        let (depth_proof, _) = joint_prove_product_from_contributions(
            &key,
            &shared["slope"].commitment,
            &shared["depth"].commitment,
            &depth_contributions,
            &quorum,
            options.threshold,
            &mut transcript,
            &mut rng,
        )?;
        let mut transcript = tagged(b"circuit:depth");
        let depth_is_slope_times_qty = verify_product(
            &key,
            &mut transcript,
            &shared["slope"].commitment,
            &qty_shared.commitment,
            &shared["depth"].commitment,
            &depth_proof,
        );

        let mut bits_proved = Map::new();
        for name in ["fits", "ok", "active"] {
            let blinding = reconstruct(&blindings[name], &quorum)?;
            let bit = reconstruct(&maker[name], &quorum)?;
            let cross_secret = blinding * (Scalar::ONE - bit);
            let cross = share_scalar(&cross_secret, &parties, options.threshold, &mut rng);
            let contributions = node_product_contributions(&shared[name], &cross, &quorum)?;
            let context = format!("circuit:{name}");
            let mut transcript = tagged(context.as_bytes());
            let (proof, _) = joint_prove_product_from_contributions(
                &key,
                &shared[name].commitment,
                &shared[name].commitment,
                &contributions,
                &quorum,
                options.threshold,
                &mut transcript,
                &mut rng,
            )?;
            let mut transcript = tagged(context.as_bytes());
            bits_proved.insert(
                name.to_string(),
                json!(verify_square_bit(
                    &key,
                    &shared[name].commitment,
                    &proof,
                    &mut transcript,
                )),
            );
        }

        let slope_value = reconstruct(&maker["slope"], &quorum)?;
        let wrong_qty_value = scalar_from_i64(options.qty + 1);
        let wrong_value_shares =
            share_scalar(&wrong_qty_value, &parties, options.threshold, &mut rng);
        let wrong_blinding_secret = Scalar::random(&mut rng);
        let wrong_blinding_shares = share_scalar(
            &wrong_blinding_secret,
            &parties,
            options.threshold,
            &mut rng,
        );
        let wrong = shared_from_maps(
            &key,
            wrong_value_shares,
            wrong_blinding_shares,
            &quorum,
            options.threshold,
        )?;
        let wrong_cross_secret = depth_blinding - slope_blinding * wrong_qty_value;
        let wrong_cross = share_scalar(&wrong_cross_secret, &parties, options.threshold, &mut rng);
        let wrong_contributions = node_product_contributions(&wrong, &wrong_cross, &quorum)?;
        let mut transcript = tagged(b"circuit:depth");
        let refused = match joint_prove_product_from_contributions(
            &key,
            &shared["slope"].commitment,
            &shared["depth"].commitment,
            &wrong_contributions,
            &quorum,
            options.threshold,
            &mut transcript,
            &mut rng,
        ) {
            Err(_) => true,
            Ok((proof, _)) => {
                let mut transcript = tagged(b"circuit:depth");
                !verify_product(
                    &key,
                    &mut transcript,
                    &shared["slope"].commitment,
                    &wrong.commitment,
                    &shared["depth"].commitment,
                    &proof,
                )
            }
        };

        let other = (index + 1) % options.makers;
        let other_blinding = Scalar::random(&mut rng);
        let other_blindings = share_scalar(&other_blinding, &parties, options.threshold, &mut rng);
        let other_depth = commitment_from_shares(
            &key,
            &scalar_wires.makers[other]["depth"],
            &other_blindings,
            &quorum,
        )?;
        let mut transcript = tagged(b"circuit:depth");
        let another_makers_depth_is_refused = !verify_product(
            &key,
            &mut transcript,
            &shared["slope"].commitment,
            &qty_shared.commitment,
            &other_depth,
            &depth_proof,
        );

        let no_node_holds_a_wire = WIRE_NAMES.iter().all(|name| {
            let opened = reconstruct(&maker[name], &quorum).expect("validated share map");
            opened == Scalar::ZERO
                || opened == Scalar::ONE
                || maker[name].values().all(|share| *share != opened)
        });
        let control_applies = slope_value != Scalar::ZERO;
        let slope = signed_i64(slope_value)?;
        println!(
            "maker {index}: commitments {commitments_from_shares_agree}, depth {depth_is_slope_times_qty}, bits {bits_proved:?}, wrong qty refused {refused} (control applies: {control_applies}), other depth refused {another_makers_depth_is_refused}"
        );
        rows.push(json!({
            "maker": index,
            "slope": slope,
            "wrong_quantity_control_applies": control_applies,
            "a_wrong_quantity_is_refused": refused,
            "another_makers_depth_is_refused": another_makers_depth_is_refused,
            "commitments_from_shares_agree": commitments_from_shares_agree,
            "depth_is_slope_times_qty": depth_is_slope_times_qty,
            "bits_proved": bits_proved,
            "no_node_holds_a_wire": no_node_holds_a_wire,
        }));
    }

    let payload = json!({
        "host": zkfmi_measure::hosts::this_host(),
        "field_bits": wires.prime.bit_length(),
        "field_matches_commitment_scalar_field": true,
        "parties": options.parties,
        "threshold": options.threshold,
        "wires_per_maker": WIRE_NAMES.len(),
        "wire_order": WIRE_NAMES,
        "runs_in_file": wires.runs_in_file,
        "winner_key": scalar_u64(winner_key)?,
        "supplied_outside_the_circuit": [
            "blindings, which the computation does not know about",
            "cross terms, which are one multiplication in a deployment",
        ],
        "rows": rows,
    });
    write_pretty_json(Some(&options.out), &payload)?;
    println!("wrote {}", options.out.display());
    Ok(())
}

struct ScalarWires {
    winner_key: ScalarShares,
    makers: Vec<BTreeMap<&'static str, ScalarShares>>,
}

fn scalar_wires(wires: &Wires) -> HarnessResult<ScalarWires> {
    Ok(ScalarWires {
        winner_key: scalar_map(&wires.winner_key)?,
        makers: wires
            .makers
            .iter()
            .map(|maker| {
                maker
                    .iter()
                    .map(|(name, shares)| Ok((*name, scalar_map(shares)?)))
                    .collect::<HarnessResult<BTreeMap<_, _>>>()
            })
            .collect::<HarnessResult<Vec<_>>>()?,
    })
}

fn scalar_map(values: &BTreeMap<usize, FieldElement>) -> HarnessResult<ScalarShares> {
    values
        .iter()
        .map(|(party, value)| Ok((*party, field_scalar(value)?)))
        .collect()
}

fn field_scalar(value: &FieldElement) -> HarnessResult<Scalar> {
    let bytes: [u8; 32] = value
        .to_bytes_le(32)?
        .try_into()
        .expect("a requested 32-byte encoding");
    Option::<Scalar>::from(Scalar::from_canonical_bytes(bytes))
        .ok_or_else(|| "a persisted share is outside the scalar field".into())
}

fn share_scalar<R: rand::RngCore + rand::CryptoRng>(
    secret: &Scalar,
    parties: &[PartyId],
    threshold: usize,
    rng: &mut R,
) -> ScalarShares {
    let points = shamir::points(parties.len());
    parties
        .iter()
        .copied()
        .zip(shamir::share(secret, threshold, &points, rng))
        .collect()
}

fn node_product_contributions(
    factor: &Shared,
    cross: &ScalarShares,
    quorum: &[PartyId],
) -> HarnessResult<Vec<ProductNodeContribution>> {
    quorum
        .iter()
        .map(|party| {
            let factor = factor
                .node_share(*party)
                .ok_or_else(|| format!("missing factor share for party {party}"))?;
            let cross_share = cross
                .get(party)
                .copied()
                .ok_or_else(|| format!("missing cross-term share for party {party}"))?;
            Ok(ProductNodeContribution::new(factor, cross_share))
        })
        .collect()
}

fn shared_from_maps(
    key: &Pedersen,
    value: ScalarShares,
    blinding: ScalarShares,
    quorum: &[PartyId],
    threshold: usize,
) -> HarnessResult<Shared> {
    let evaluations = value
        .iter()
        .map(|(party, share)| (*party, key.commit(share, &blinding[party])))
        .collect::<BTreeMap<_, _>>();
    Ok(Shared {
        commitment: commitment_from_shares(key, &value, &blinding, quorum)?,
        coefficient_commitments: coefficient_commitments_from_evaluations(&evaluations, threshold)?,
        value,
        blinding,
    })
}

fn wire_from_shares<R: rand::RngCore + rand::CryptoRng>(
    key: &Pedersen,
    value: &ScalarShares,
    parties: &[PartyId],
    quorum: &[PartyId],
    threshold: usize,
    rng: &mut R,
) -> HarnessResult<(Shared, ScalarShares)> {
    let blinding = share_scalar(&Scalar::random(rng), parties, threshold, rng);
    Ok((
        shared_from_maps(key, value.clone(), blinding.clone(), quorum, threshold)?,
        blinding,
    ))
}

fn reconstruct(shares: &ScalarShares, quorum: &[PartyId]) -> HarnessResult<Scalar> {
    let values = quorum
        .iter()
        .map(|party| {
            shares
                .get(party)
                .copied()
                .ok_or_else(|| format!("missing share for party {party}").into())
        })
        .collect::<HarnessResult<Vec<_>>>()?;
    let points = quorum
        .iter()
        .map(|party| Scalar::from(*party as u64))
        .collect::<Vec<_>>();
    Ok(shamir::reconstruct(&points, &values))
}

fn scalar_from_i64(value: i64) -> Scalar {
    if value < 0 {
        -Scalar::from(value.unsigned_abs())
    } else {
        Scalar::from(value as u64)
    }
}

fn scalar_u64(value: Scalar) -> HarnessResult<u64> {
    let bytes = value.to_bytes();
    if bytes[8..].iter().any(|byte| *byte != 0) {
        return Err("a circuit result does not fit u64".into());
    }
    Ok(u64::from_le_bytes(bytes[..8].try_into().unwrap()))
}

fn signed_i64(value: Scalar) -> HarnessResult<i64> {
    if let Ok(positive) = scalar_u64(value) {
        return i64::try_from(positive)
            .map_err(|_| "a positive circuit wire does not fit i64".into());
    }
    let magnitude = scalar_u64(-value)?;
    i64::try_from(magnitude)
        .map(|value| -value)
        .map_err(|_| "a negative circuit wire does not fit i64".into())
}

fn tagged(context: &[u8]) -> Transcript {
    let mut transcript = Transcript::new(b"qomm:circuit-bound:v1");
    transcript.append_message(b"context", context);
    transcript
}

fn parse_args() -> HarnessResult<Options> {
    let mut persistence = None;
    let mut options = Options {
        persistence: PathBuf::new(),
        out: repo_root().join("artifacts/circuit_bound_proof.json"),
        parties: 7,
        threshold: 2,
        makers: 4,
        qty: 100,
    };
    let mut args = std::env::args_os().skip(1);
    while let Some(flag) = args.next() {
        match flag.to_str() {
            Some("--persistence") => {
                persistence = Some(PathBuf::from(next(&mut args, "--persistence")?))
            }
            Some("--out") => options.out = PathBuf::from(next(&mut args, "--out")?),
            Some("--parties") => {
                options.parties = parse_value(next(&mut args, "--parties")?, "--parties")?
            }
            Some("--threshold") => {
                options.threshold = parse_value(next(&mut args, "--threshold")?, "--threshold")?
            }
            Some("--makers") => {
                options.makers = parse_value(next(&mut args, "--makers")?, "--makers")?
            }
            Some("--qty") => options.qty = parse_value(next(&mut args, "--qty")?, "--qty")?,
            Some("-h" | "--help") => {
                println!("usage: run_circuit_bound_proof --persistence PATH [--out PATH] [--parties N] [--threshold N] [--makers N] [--qty N]");
                std::process::exit(0);
            }
            _ => return Err(format!("unknown argument {}", flag.to_string_lossy()).into()),
        }
    }
    options.persistence = persistence.ok_or("--persistence is required")?;
    Ok(options)
}

fn next(args: &mut impl Iterator<Item = OsString>, flag: &str) -> HarnessResult<OsString> {
    args.next()
        .ok_or_else(|| format!("argument {flag} expects one value").into())
}
